### PR TITLE
Add mac packaged runtime and beta release flow

### PR DESCRIPTION
**Pull request by PerishCode**
## Summary
- Add the mac-first `tools-pack` packaging/control-plane flow for build, install, start, stop, logs, uninstall, and cleanup.
- Add the packaged Electron runtime that launches daemon/web sidecars with namespace-scoped paths and `od://` entry routing.
- Add beta release automation for signed/unsigned mac artifacts, portable configs, update feed assets, and cleaned mac install visuals.

## Validation
- `pnpm install`
- `pnpm typecheck`
- `pnpm test`
- `pnpm build`
- `tools-dev` smoke: start/status/inspect/check/logs/stop with namespace `merge-dev-smoke`
- `tools-pack` smoke: build/install/start/inspect/logs/stop/uninstall/cleanup with namespace `merge-pack-smoke`
- Release workflow: unsigned beta4 succeeded; signed beta5 succeeded with app/ZIP codesign, Gatekeeper, and stapler validation

## Notes
- Signed app and ZIP notarization are validated. The signed DMG container itself still needs signing/stapling follow-up before calling the DMG container flow fully complete.

### CODE REVIEW

File: apps/daemon/src/server.ts [hunk 1]
```diff
@@ -79,13 +79,38 @@ export function resolveProjectRoot(moduleDir: string): string {
 }
 
 const PROJECT_ROOT = resolveProjectRoot(__dirname);
+const RESOURCE_ROOT_ENV = 'OD_RESOURCE_ROOT';
+
+function resolveDaemonResourceRoot() {
+  const configured = process.env[RESOURCE_ROOT_ENV];
+  return configured && configured.length > 0 ? path.resolve(configured) : null;
+}
+
```
Comment (lefarcen) [inline apps/daemon/src/server.ts:121]: 🔒 **P1 Path traversal risk:** `process.env.OD_RESOURCE_ROOT` is resolved without validation. A malicious config could set `OD_RESOURCE_ROOT=../../../../etc` to serve arbitrary filesystem paths via `/skills`, `/design-systems`, or `/frames` routes.

**Fix:** Add allowlist validation requiring resource root to be under `PROJECT_ROOT` or `process.resourcesPath`:
```typescript
function resolveDaemonResourceRoot() {
  const configured = process.env[RESOURCE_ROOT_ENV];
  if (!configured) return null;
  const resolved = path.resolve(configured);
  const safeBases = [PROJECT_ROOT, process.resourcesPath].filter(Boolean);
  if (!safeBases.some(base => resolved.startsWith(base))) {
    throw new Error(`OD_RESOURCE_ROOT must be under workspace or app bundle`);
  }
  return resolved;
}
```

File: apps/web/sidecar/server.ts [hunk 2]
```diff
@@ -64,6 +72,53 @@ function parsePort(value: string | undefined): number {
...
+function shouldProxyToDaemon(requestUrl: string | undefined): boolean {
+  if (requestUrl == null) return false;
+  const pathname = new URL(requestUrl, `http://${HOST}`).pathname;
+  return pathname === "/api" || pathname.startsWith("/api/") || pathname === "/artifacts" || pathname.startsWith("/artifacts/") || pathname === "/frames" || pathname.startsWith("/frames/");
+}
+
+async function proxyToDaemon(
+  daemonOrigin: string,
+  request: IncomingMessage,
+  response: ServerResponse,
```
Comment (lefarcen) [inline apps/web/sidecar/server.ts:112]: 🔒 **P1 SSRF potential:** `proxyToDaemon()` constructs `new URL(request.url, daemonOrigin)` without validating `request.url`. If the web sidecar is exposed to external requests (future feature or misconfigured proxy), an attacker could inject `request.url = "http://internal-service/admin"` to SSRF internal services.

**Fix:** Validate pathname against allowlist before proxying:
```typescript
function shouldProxyToDaemon(requestUrl: string | undefined): boolean {
  if (!requestUrl) return false;
  const pathname = new URL(requestUrl, `http://${HOST}`).pathname;
  return pathname === "/api" || pathname.startsWith("/api/") ||
         pathname === "/artifacts" || pathname.startsWith("/artifacts/") ||
         pathname === "/frames" || pathname.startsWith("/frames/");
}
// Then reject in proxyToDaemon if pathname doesn't match
```

File: apps/packaged/src/sidecars.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    source: options.runtime.source,
+  } satisfies SidecarStamp;
+  const logHandle = await openLog(logPathFor(options.paths, options.app));
+  const childEnv = createSidecarLaunchEnv({
+    base: options.paths.runtimeRoot,
+    contract: OPEN_DESIGN_SIDECAR_CONTRACT,
+    extraEnv: {
+      ...process.env,
+      ...options.env,
+      NODE_ENV: "production",
```
Comment (lefarcen) [inline apps/packaged/src/sidecars.ts:160]: **P2 Env leakage:** `extraEnv: { ...process.env, ...options.env }` inherits the packaged desktop's full environment, which may include user secrets (e.g., `SSH_AUTH_SOCK`, `GITHUB_TOKEN` if desktop was launched from a dev shell). The daemon/web sidecars don't need these.

**Fix:** Use an explicit allowlist:
```typescript
const allowedEnvKeys = ['PATH', 'HOME', 'USER', 'LANG', 'TMPDIR'];
const baseEnv = Object.fromEntries(
  allowedEnvKeys.map(key => [key, process.env[key]]).filter(([_, v]) => v != null)
);
const childEnv = createSidecarLaunchEnv({ extraEnv: { ...baseEnv, ...options.env, NODE_ENV: "production", ... }, ... });
```

File: scripts/release-beta.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  if (!stableVersionPattern.test(packageJson.version)) {
+    fail(`apps/packaged/package.json version must be a stable x.y.z base version; got ${packageJson.version}`);
+  }
+
+  return packageJson.version;
+}
+
+async function fetchReleases(repository: string): Promise<GitHubRelease[]> {
```
Comment (lefarcen) [inline scripts/release-beta.ts:134]: **P2 Pagination missing:** `gh api ... ?per_page=100` fetches only the first page. If the repo has >100 releases, the script won't see older betas, potentially colliding beta numbers or missing the true latest stable.

**Fix:** Implement pagination:
```typescript
async function fetchReleases(repository: string): Promise<GitHubRelease[]> {
  let page = 1;
  const allReleases: GitHubRelease[] = [];
  while (true) {
    const { stdout } = await execFile("gh", ["api", `repos/${repository}/releases?per_page=100&page=${page}`]);
    const batch = JSON.parse(stdout) as GitHubRelease[];
    if (batch.length === 0) break;
    allReleases.push(...batch);
    page += 1;
  }
  return allReleases;
}
```

File: tools/pack/src/mac.ts [hunk 5]
```diff
@@ -0,0 +1,1001 @@
...
+    return {
+      fallback: { ...fallback, reason: "marker-namespace-root-mismatch" },
+      rootPid: null,
+    };
+  }
+
+  const processes = await listProcessSnapshots();
+  const processInfo = processes.find((entry) => entry.pid === marker.pid) ?? null;
+  if (processInfo == null) {
+    return {
```
Comment (lefarcen) [inline tools/pack/src/mac.ts:726]: **P3 Stale marker cleanup:** If `stopProcesses()` succeeds but the marker file remains (filesystem error), subsequent `start` might refuse to run. Add explicit cleanup:
```typescript
if (stopped.remainingPids.length === 0) {
  await rm(desktopIdentityPath(config), { force: true }).catch(() => undefined);
}
```

File: apps/AGENTS.md [hunk 6]
```diff
@@ -20,10 +21,12 @@ Follow the root `AGENTS.md` first. This file only records module-level boundarie
 - App business layers must not import sidecar packages or branch on `runtime.mode`, `namespace`, `ipc`, or `source`.
 - Keep sidecar awareness in `apps/<app>/sidecar` or the desktop sidecar entry wrapper.
 
-## Inactive app directories
+## Packaged runtime
 
 - `apps/nextjs` has been removed; do not restore it.
-- `apps/packaged` is a minimal placeholder for future packaged app assembly. Do not add a package manifest, runtime code, or lifecycle script there in this round.
+- Packaged web uses Next.js SSR through the web sidecar; do not put Next output under daemon `OD_RESOURCE_ROOT`.
```
Comment (lefarcen) [inline apps/AGENTS.md:27]: **P2 Reasoning gap (Lens B):** Doc claims "ports are transient", but `apps/packaged/src/sidecars.ts:212` still passes `OD_DAEMON_PORT` to the web sidecar, creating a dependency on daemon's URL. Why not Unix domain sockets?

**Suggest adding:**
> **Why not Unix sockets for daemon↔web?** Next.js dev server and SSR assume HTTP origins for rewrites/proxies. Switching to UDS would require patching Next.js internals. We accept the daemon-port-to-web-env dependency as pragmatic; the key invariant is that *data/log/runtime paths* don't embed ports.

File: tools/pack/README.md [hunk 7]
```diff
@@ -1,5 +1,55 @@
...
+- `data/` is the daemon-managed data root passed to the daemon through the packaged sidecar launch environment.
+- `logs/` contains packaged process logs for `desktop`, `web`, and `daemon`.
+- `runtime/` is the sidecar runtime base used by the packaged desktop/web/daemon process group.
+- `cache/` is reserved for namespace-local packaged cache state.
+- `user-data/` is the Electron/Chromium `userData` root, with `user-data/session/` used for `sessionData`.
+
+Finder/manual launches cannot carry argv stamps on the root desktop process. To keep process fallback safe,
+`apps/packaged` writes `runtime/desktop-root.json` with the desktop stamp, PID, executable path, app path, and log path.
+`tools-pack mac stop` trusts that marker only when namespace/stamp/PID/command validation passes; otherwise it reports the
+unmanaged/not-owned reason instead of killing unknown processes.
```
Comment (lefarcen) [inline tools/pack/README.md:33]: **P2 Missing failure modes (Lens B):** README describes the marker file but doesn't explain what happens if marker is stale/corrupted/mismatched. The code handles it (see `resolveDesktopRootIdentityFallback`), but users should know `stop` won't kill unrelated processes.

**Suggest adding a section:**
> ### tools-pack stop validation
> - If marker is absent → reports "not-running"
> - If marker PID is reused by unrelated process → reports "unmanaged"
> - If marker namespace/stamp mismatches → reports "unmanaged"
> 
> This ensures `stop` never kills processes outside the current namespace.

**Review by lefarcen — CHANGES_REQUESTED** [review]
Hey @PerishCode, thanks for the PR! 🎉

This is a significant infrastructure addition — the packaged runtime architecture is solid, and the namespace-scoped path model is a good long-term foundation. The beta release automation is thorough.

Found **2 critical security issues** (daemon path traversal, web proxy SSRF potential) that need fixes before merge, plus several P2 design gaps (beta pagination, env leakage, reasoning completeness in AGENTS.md). See inline comments below.

**High-level observations:**
- ✅ Good: Desktop identity marker prevents accidental process kills
- ✅ Good: tools-pack stop graceful→forceful escalation
- ⚠️ Needs fix: OD_RESOURCE_ROOT lacks path validation (P1)
- ⚠️ Needs fix: Web daemon proxy allows arbitrary URLs (P1)
- ⚠️ Needs improvement: Beta release script only fetches first 100 releases (P2)

Details inline.

**Comment by PerishCode** [comment]
Addressed the review feedback in a10167a.

Summary:
- Hardened daemon `OD_RESOURCE_ROOT` with containment checks for workspace/app resources and added coverage.
- Pinned web sidecar daemon proxy targets to the configured daemon origin and added absolute-URL regression coverage.
- Switched packaged daemon/web sidecars to an explicit child env allowlist while preserving packaged PATH augmentation.
- Paginated beta release lookup beyond the first 100 releases.
- Cleaned stale desktop identity markers after successful stop/no-managed-process states.
- Documented the daemon↔web HTTP-port tradeoff and `tools-pack mac stop` marker failure modes.

Validation:
- `pnpm --filter @open-design/daemon test`
- `pnpm --filter @open-design/daemon typecheck`
- `pnpm --filter @open-design/web test`
- `pnpm --filter @open-design/web typecheck`
- `pnpm --filter @open-design/packaged typecheck`
- `pnpm --filter @open-design/tools-pack typecheck`
- `pnpm --filter @open-design/tools-pack build`
- `pnpm --filter @open-design/packaged build`
- `pnpm typecheck && pnpm test && pnpm build`
- `tools-pack mac build/start/logs/stop/cleanup --namespace review-fix-smoke --to app`, including a second `stop` returning `marker-not-found` after marker cleanup.

File: apps/daemon/src/server.ts (hunk 1 above)
Comment (PerishCode) [reply apps/daemon/src/server.ts:121]: Fixed in a10167a. `OD_RESOURCE_ROOT` now resolves through a containment check against the workspace root or app Resources directory; the packaged Node sidecar also infers the macOS Resources path when `process.resourcesPath` is unavailable. Added daemon tests for allowed, equal, and outside roots.

File: apps/web/sidecar/server.ts (hunk 2 above)
Comment (PerishCode) [reply apps/web/sidecar/server.ts:112]: Fixed in a10167a. The web sidecar now validates the pathname and rebuilds the proxy target from pathname/search on the configured daemon origin, so absolute request URLs cannot replace the destination origin. Added regression coverage.

File: tools/pack/README.md (hunk 7 above)
Comment (PerishCode) [reply tools/pack/README.md:33]: Added in a10167a. `tools/pack/README.md` now documents stop validation for absent, stale, reused, and mismatched markers and states that stop will not kill outside the current namespace.

File: tools/pack/src/mac.ts (hunk 5 above)
Comment (PerishCode) [reply tools/pack/src/mac.ts:726]: Fixed in a10167a. `stopPackedMacApp` removes `desktop-root.json` after successful no-managed-process states and after force stop leaves no remaining PIDs; unmanaged mismatches are preserved. Smoke verified a second stop returns `marker-not-found`.

File: apps/packaged/src/sidecars.ts (hunk 3 above)
Comment (PerishCode) [reply apps/packaged/src/sidecars.ts:160]: Fixed in a10167a. Packaged daemon/web sidecars now inherit only a small allowlist (`HOME`, `LANG`, `LC_ALL`, `LOGNAME`, `TMPDIR`, `USER`) plus required packaged env and the augmented `PATH`.

File: scripts/release-beta.ts (hunk 4 above)
Comment (PerishCode) [reply scripts/release-beta.ts:134]: Fixed in a10167a. `fetchReleases` now pages through `per_page=100&page=N` until GitHub returns an empty page.

File: apps/AGENTS.md (hunk 6 above)
Comment (PerishCode) [reply apps/AGENTS.md:27]: Added in a10167a. `apps/AGENTS.md` now documents why packaged daemon↔web traffic stays on HTTP origin/port and clarifies that the invariant is path independence from ports.

**Review by lefarcen — APPROVED** [review]
Hey @PerishCode! 🎉

All P1/P2/P3 items from the previous review have been addressed. The fixes are excellent:

**P1 security fixes (both complete):**
- ✅ Daemon path traversal: `isPathWithin()` validation + safe bases check
- ✅ Web proxy SSRF: pathname allowlist + validation before URL construction

**P2 improvements:**
- ✅ Env leakage: `PACKAGED_CHILD_ENV_ALLOWLIST` filters inherited env
- ✅ Beta pagination: loop-until-empty pattern implemented
- ✅ AGENTS.md reasoning gap: HTTP-vs-UDS tradeoff documented
- ✅ tools/pack README: failure modes section added

**P3:**
- ✅ Marker cleanup: stale file removed on successful stop

**Bonus:** Added comprehensive tests (`server-paths.test.ts`, `sidecar-proxy.test.ts`) covering path validation, SSRF prevention, and edge cases — significantly increases confidence in the security fixes.

Great work on the thorough response. LGTM! 🚀

**Review by lefarcen — APPROVED** [review]
